### ISSUE TITLE
planning: Train our own Quantizer for multilingual speech to work with Whisper Encoder

### ISSUE THREAD

**Issue by hahuyhoang411** [janhq/ichigo#146]
## Goal
Experiment on WhisperVQ model for better result on multilingual. Hypothesis the current codebook is only 512 which is a small space to compress the multilingual capability.

## Learning Goals 

- Train our own Quantizer
- Gain more understanding of quantization 
    - [WhisperSpeech](https://github.com/lucidrains/vector-quantize-pytorch) currently uses ResidualVQ
    - Can we explore other [VQ methods](https://github.com/collabora/WhisperSpeech/tree/main?tab=readme-ov-file#encodec-for-modeling-acoustic-tokens) (e.g. SimVQ, Lookup-free, etc)
- Quantizer training leads to more reusable efforts
    - Can quantize images, etc

<img width="1268" alt="Screenshot 2024-11-27 at 15 25 51" src="https://github.com/user-attachments/assets/43eb20fd-a008-430f-aecd-9d6ebc732732">

## Tasklist
- [x] Check the [WhisperSpeech](https://github.com/collabora/WhisperSpeech) training codebase: 
- [x] Experiment with model parameters (codebook, batchsize, data pipeline, etc)
- [ ] janhq/ichigo#144 
- [ ] Draft research report

## Experiments

- We expand our quantizer codebook size from 512 (original weight) to 1024 for better capturing multilingual.
  - [Discussion](https://github.com/janhq/ichigo/issues/144), [Implementation](https://github.com/janhq/WhisperSpeech/pull/7)
- The first issue we encountered while training the quantizer model was a high KL loss. Detailed debugging notes are available in our worklog: [GitHub Issue janhq/ichigo#144](https://github.com/janhq/ichigo/issues/144).

Run ID | Date (dd/mm/yy) | Model Config | Dataset | Learning Rate | Batch Size | Steps | Total Loss | CE Loss | KL Loss | Commit Loss | Codebook Utilization | Runtime | Hardware | Note
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
expand-exp1 | 02/12/24 | codes=1024 (random init), max_token=200 | Bud500 | 1e-3 | 8 | 9999 | 15.83308 | 0.76369 | 15.06781 | 0.0015839 | 86% | 1h 30m | 1xA6000 | small batch size is not optimal
expand-exp2 | 02/12/24 | codes=1024 (random init), max_token=200 | Bud500 | 1e-3 | 42 | 1689 | 15.96285 | 0.83534 | 15.12557 | 0.001936 | 95% | 1h 5m | 1xA6000 | random init got higher codebook utils
expand-exp3 | 02/12/24 | codes=1024 (avg init), max_token=200 | Bud500 | 1e-3 | 42 | 4999 | 15.39613 | 0.80295 | 14.59104 | 0.0021388 | 84% | 3h 21m | 1xA6000 |  
kl-exp1 | 03/12/24 | codes=1024, KL=5,max_token=200 | Bud500 | 1e-3 | 42 | 782 | 78.53279 | 1.11499 | 15.48318 | 0.001883 | 83% | 31m 1s | 1xA6000 | change KL factor but still high KL loss
kl-exp2 | 03/12/24 | codes=1024, KL=2,max_token=200 | Bud500 | 1e-3 | 42 | 2924 | 30.53683 | 0.91597 | 14.80941 | 0.002393 | 85% | 1h 57m | 1xA6000 | change KL factor but still high KL loss
kl-exp3 | 03/12/24 | codes=1024, KL=1.5,max_token=200 | Bud500 | 1e-3 | 42 | 2939 | 23.34618 | 0.84657 | 14.99841 | 0.001992 | 86% | 1h 58m | 1xA6000 | change KL factor but still high KL loss
kl-exp4 | 03/12/24 | codes=1024, KL=3,max_token=200 | Bud500 | 1e-3 | 42 | 3726 | 45.72793 | 1.0146 | 14.90375 | 0.002078 | 84% | 2h 43m | 1xA6000 | change KL factor but still high KL loss
mask-logit-exp | 03/12/24 | codes=1024, KL=1, max_token=200 | Bud500 | 1e-3 | 42 | 2335 | 0.60911 | 0.29703 | 0.3107 | 0.0013738 | 86% | N/A | 1xA6000 | missing loss calculation (deprecated implementation)
expand-exp4 | 04/12/24 | codes=1024 (random init), max_token=50 | Bud500 | 1e-3 | 42 | 15099 | 4.71717 | 0.7786 | 3.93544 | 0.003314 | 96% | 7h 43m | 1xA6000 | decrease tokens cap = reduce pad tokesn = reduce KL loss
expand-exp5 | 04/12/24 | codes=1024 (random init), max_token=20 | Bud500 | 1e-3 | 42 | 15099 | 2.75478 | 0.82511 | 1.92655 | 0.0031208 | 94% | 7h 32m | 1xA6000 | tokens cap = max tokens of data (20) = optimal KL loss
